documentation/specs/MultithreadedMSBuild/thread-safe-tasks.md

```diff
@@ -0,0 +1,171 @@
+
+To enable this multithreaded execution model, we introduce the `IThreadSafeTask` interface that tasks can implement to declare their thread-safety capabilities. Tasks implementing this interface must avoid using APIs that modify or depend on global process state, as this could cause conflicts when multiple tasks execute concurrently. See [Thread-Safe Tasks API Reference](thread-safe-tasks-api-reference.md) for detailed guidelines. At the same time, the `IThreadSafeTask` interface provides access to the `ExecutionContext` property that allows safe access to global process state. For example, task authors should use `ExecutionContext.GetAbsolutePath(relativePath)` instead of the standard `Path.GetFullPath(relativePath)` to ensure correct path resolution.
+
+## Design
+
+```csharp
+public interface IThreadSafeTask : ITask
+{
+    ITaskExecutionContext ExecutionContext { get; set; }
+}
+```
+
+**Note:** Consider backporting the `IThreadSafeTask` interface to the 17.14 branch to allow a graceful fail when tasks attempt to use it.
+
+The execution context provides essential methods for thread-safe task execution, replacing direct access to global process state:
+
+```csharp
+public interface ITaskExecutionContext
+{
+    AbsolutePath CurrentDirectory { get; set; }
+
+    AbsolutePath GetAbsolutePath(string path);
+    
+    string? GetEnvironmentVariable(string name);
+    IReadOnlyDictionary<string, string> GetEnvironmentVariables();
+    void SetEnvironmentVariable(string name, string? value);
+
+    ProcessStartInfo GetProcessStartInfo();
+    Process StartProcess(ProcessStartInfo startInfo);
+    Process StartProcess(string fileName);
+    Process StartProcess(string fileName, IEnumerable<string> arguments);
+}
+```
+
+**Note:** The `ITaskExecutionContext` will not be thread-safe for performance reasons. Task authors who spawn multiple threads within their task implementation must provide their own synchronization when accessing the execution context from multiple threads. However, each thread node has its own isolated context object provided to the tasks, so task authors do not need to worry about synchronization with other tasks running concurrently in different thread nodes.
+
+To help task authors avoid thread-safety issues related to path handling, we introduce `AbsolutePath` and `RelativePath` classes that are implicitly convertible to string. 
+
+```csharp
+public class AbsolutePath
+{
+    public string Path { get; }
+    
+    // Will be banned for use in tasks by analyzers
+    public AbsolutePath(string path) { }
+    
+    public AbsolutePath(string path, AbsolutePath basePath) { }
+    
+    public static implicit operator string(AbsolutePath path) { }
```

Comment:
Implicit conversions can cause confusion. If they don't provide a large degree of convenience for end users, I'd consider dropping them.

Reply:
this is indeed the case where I think some large degree of convenience could be - if you consider that AbsolutePath is really a verified string (verified that it contains indeed absolute path), you would like to use it as such